src/uwtools/rocoto.py

```diff
@@ -143,3 +65,252 @@ def validate_rocoto_xml(input_xml: OptionalPath) -> bool:
+    def _config_validate(self, config_file: OptionalPath) -> None:
+        """
+        Validate the given YAML config.
+
+        :param config_file: Path to the YAML config (defaults to stdin).
+        """
+        if not validate_yaml(
+            config_file=config_file, schema_file=resource_pathobj("rocoto.jsonschema")
+        ):
+            raise UWConfigError("YAML validation errors identified in %s" % config_file)
+
+    def _add_metatask(self, e: Element, config: dict, taskname: str) -> None:
+        """
+        Add a <metatask> element to the <workflow>.
+
+        :param e: The parent element to add the new element to.
+        :param config: Configuration data for this element.
+        :param taskname: The name of the metatask being defined.
+        """
+        e = SubElement(e, STR.metatask, name=taskname)
+        for key, val in config.items():
+            if key.startswith(STR.metatask):
+                self._add_metatask(e, val, self._tag_name(key)[1])
+            elif key.startswith(STR.task):
+                self._add_task(e, val, self._tag_name(key)[1])
+            elif key == STR.var:
+                for name, value in val.items():
+                    SubElement(e, STR.var, name=name).text = value
+
+    def _add_task(self, e: Element, config: dict, taskname: str) -> None:
+        """
+        Add a <task> element to the <workflow>.
+
+        :param e: The parent element to add the new element to.
+        :param config: Configuration data for this element.
+        :param taskname: The name of the task being defined.
+        """
+        e = SubElement(e, STR.task, name=taskname)
+        self._set_attrs(e, config)
+        # These elements are simple enough to not require their own methods:
+        for name in (STR.account, STR.command, STR.nodes, STR.walltime):
+            if name in config:
+                SubElement(e, name).text = config[name]
+        # The job name, if unspecified, defaults to the task name.
+        SubElement(e, STR.jobname).text = config.get(STR.jobname, taskname)
+        # Add any dependencies:
+        if STR.dependency in config:
+            self._add_task_dependency(e, config[STR.dependency])
+        # Add any environment-variable entities:
+        for name, value in config.get(STR.envars, {}).items():
+            self._add_task_envar(e, name, value)
+
+    def _add_task_dependency(self, e: Element, config: dict) -> None:
+        """
+        Add a <dependency> element to the <task>.
+
+        :param e: The parent element to add the new element to.
+        :param config: Configuration data for this element.
+        """
+        e = SubElement(e, STR.dependency)
+        for key, block in config.items():
+            tag, _ = self._tag_name(key)
+            if tag == STR.taskdep:
+                d = SubElement(e, STR.taskdep)
+                self._set_attrs(d, block)
+            else:
+                raise UWConfigError("Unhandled dependency type %s" % tag)
```

Comment:
Dependencies are much more complicated in general than what we have in our tests. We should consider adding that in our tests as part of the follow-up. This will likely end up being a recursive call to ensure nested dependencies are fleshed out. It gets more complicated in that there are 4 (I _think_) general categories of how dependencies are built: 

* Some have "text" attributes like `<left>"foo"</left>`
* Some have element attributes like `<taskdep task="foo"/>`
* Some have both text attributes and element attributes like `<datadep age=05:00>/some/path/to/data</datadep>`
* Some have subelements like `<or>`, `<streq>`, etc., but no element attributes or text attributes.

Reply:
Yes, implementation of the rest of the dependency complexity should be done in lock-step with tests that exercise all those possibilities.